gh-85602: Add docs for generating C function signatures (#21673)
python/cpython: Doc/extending/extending.rst

Use :c:func:`PyArg_ParseTupleAndKeywords` to parse the arguments to such a
function.

> The fourth entry is the function's docstring. Note the
> ``"system(command)\n--\n\n"`` at the beginning. Since C extension functions
> take a fixed number of arguments---e.g. two for ``METH_VARARGS``---the Python
> interpreter has no way of knowing the function's signature until the function
> is called. However, by making the docstring start with something of the form
> ``"myfunc(arg1, arg2='my_default', arg3=17)\n--\n\n"``, the signature will be
> constructed and the remainder of the docstring will be used as the function's
> ``__doc__`` attribute. This can be useful for generating nice messages with the
> :func:`help` function or for getting the signature in object form with the
> :func:`inspect.signature` function.

Reviewer: ```suggestion
The fourth entry is the function's docstring. Note the
``"system(command)\n--\n\n"`` at the beginning. Since C extension functions
take a fixed number of arguments (for instance, two for ``METH_VARARGS``) the Python
interpreter has no way of knowing the function's signature until the function
is called. However, by making the docstring start with something of the form
``"myfunc(arg1, arg2='my_default', arg3=17)\n--\n\n"``, the signature will be
constructed and the remainder of the docstring will be used as the function's
``__doc__`` attribute. This is typically useful for generating nice messages with
:func:`help` or for getting the signature object with :func:`inspect.signature`.
```

Use :c:func:`PyArg_ParseTupleAndKeywords` to parse the arguments to such a
function.

> The fourth entry is the function's docstring. Note the ``"system(command)\n--\n\n"``

Reviewer 2: Please limit lines to 80 characters (https://devguide.python.org/documenting/#use-of-whitespace).

at the beginning. Since C extension functions take a fixed number of arguments---
e.g. 2 for ``METH_VARARGS``---the Python interpreter has no way of knowing the
function's signature until the function is called. However, by making the docstring
start with something of the form
``"myfunc(arg1, arg2='my_default', arg3=17)\n--\n\n"``, the signature will be
constructed and the remainder of the docstring will be used as the function's
``__doc__`` attribute. This can be useful for generating nice messages with the

> ``help`` function or ``Signature`` objects with the :mod:`inspect` module.

Reviewer 2: ```rst
:func:`help`
```

```rst
:class:`~inspect.Signature`
```

Author: Thanks for pointing this out!
